haystack/preview/document_stores/memory/document_store.py

```diff
@@ -142,3 +165,72 @@ def delete_documents(self, document_ids: List[str]) -> None:
             if not doc_id in self.storage.keys():
                 raise MissingDocumentError(f"ID '{doc_id}' not found, cannot delete it.")
             del self.storage[doc_id]
+
+    def bm25_retrieval(
+        self, query: str, filters: Optional[Dict[str, Any]] = None, top_k: int = 10, scale_score: bool = True
+    ) -> List[Document]:
+        """
+        Retrieves documents that are most relevant to the query using BM25 algorithm.
+
+        :param query: The query string.
+        :param filters: A dictionary with filters to narrow down the search space.
+        :param top_k: The number of top documents to retrieve. Default is 10.
+        :param scale_score: Whether to scale the scores of the retrieved documents. Default is True.
+        :return: A list of the top 'k' documents most relevant to the query.
+        """
+        if not query:
+            raise ValueError("Query should be a non-empty string")
+
+        # Get all documents that match the user's filters AND are either 'table' or 'text'.
+        # Raises an exception if the user was trying to include other content types.
+        if filters and "content_type" in filters:
+            content_types = filters["content_type"]
+            if isinstance(content_types, str):
+                content_types = [content_types]
+            if any(type_ not in ["text", "table"] for type_ in content_types):
+                raise ValueError(
+                    "MemoryDocumentStore can do BM25 retrieval on no other document type than text or table."
+                )
+        else:
+            filters = filters or {}
+            filters = {**filters, "content_type": ["text", "table"]}
+        all_documents = self.filter_documents(filters=filters)
+
+        # Lowercase all documents
+        lower_case_documents = []
```

Comment:
@ZanSara @silvanocerza, the design decisions we will make for our previous [discussion](https://github.com/deepset-ai/haystack/pull/5151/files#r1243259423) influence how we implement these components. Specifically, if we cannot ensure that the input fields from another component (here, MemoryRetriever.Input) passed to the document store (this method) will never be None, then we must introduce safeguards for None similar to filters. This applies even for primitive parameters top_k and scale_score.

Reply:
https://github.com/deepset-ai/canals/issues/33